contributors/design-proposals/autoscaling/hpa-external-metrics.md

```diff
@@ -0,0 +1,144 @@
+        ExternalMetricSourceType MetricSourceType = "External"
+)
+
+// ObjectMetricSource indicates how to scale on a metric describing a
+// Kubernetes object (for example, hits-per-second on an Ingress object).
+type ObjectMetricSource struct {
+        // target is the described Kubernetes object.
+        Target CrossVersionObjectReference `json:"target" protobuf:"bytes,1,name=target"`
+
+        // metricName is the name of the metric in question.
+        MetricName string `json:"metricName" protobuf:"bytes,2,name=metricName"`
+        // TargetValue is the target value of the metric (as a quantity).
+        // Mutually exclusive with TargetAverageValue.
+        TargetValue *resource.Quantity `json:"targetValue,omitempty" protobuf:"bytes,3,opt,name=targetValue"`
+        // TargetAverageValue is the target per-pod value of global metric.
+        // Mutually exclusive with TargetValue.
+        TargetAverageValue *resource.Quantity `json:"targetAverageValue,omitempty" protobuf="bytes,4,opt,name=targetAverageValue"`
+}
+
+// ExternalMetricSource indicates how to scale on a metric coming from
+// outside of Kubernetes cluster (for example length of queue in cloud
+// messaging service, or QPS from loadbalancer running outside of cluster).
+type ExternalMetricSource struct {
+        // MetricName is the name of a metric used for autoscaling in
+        // metric system.
+        MetricName string `json:"metricName" protobuf:"bytes,1,name=metricName"`
+
+        // MetricSelector is used to identify a specific time series
+        // within a given metric.
+        MetricSelector metav1.LabelSelector `json:"metricSelector" protobuf:"bytes,2,name=metricSelector"`
+
+        // TargetValue is the target value of the metric (as a quantity).
+        // Mutually exclusive with TargetAverageValue.
+        TargetValue *resource.Quantity `json:"targetValue,omitempty" protobuf:"bytes,3,opt,name=targetValue"`
+        // TargetAverageValue is the target per-pod value (as a quantity) of global metric.
+        // Mutually exclusive with TargetValue.
+        TargetAverageValue *resource.Quantity `json:"targetAverageValue,omitempty" protobuf="bytes,4,opt,name=targetAverageValue"`
+}
+
+// ExternalMetricStatus indicates the current value of a global metric
+// coming from outside of Kubernetes.
+type ExternalMetricStatus struct {
+        // MetricName is the name of a metric used for autoscaling in
+        // metric system.
+        MetricName string `json:"metricName" protobuf:"bytes,1,name=metricName"`
+
+        // MetricSelector is used to identify a specific time series
+        // within a given metric.
+        MetricSelector metav1.LabelSelector `json:"metricSelector" protobuf:"bytes,2,name=metricSelector"`
+
+        // CurrentValue is the current value of the metric (as a quantity)
+        CurrentValue resource.Quantity `json:"currentValue" protobuf:"bytes,3,name=currentValue"`
```

Comment:
This also needs `currentAverageValue *resource.Quantity` as an optional field.

Reply:
Added.